ENH: Adding Modified Rodrigues Parameters to the Rotation class

nbelakovski opened the pull request:
#### Reference issue
N/A

#### What does this implement/fix?
The implements 2 new methods in the Rotation class of the spatial/transform package, as_mrp and from_mrp, which allow the class to accept input and provide output in the form of Modified Rodrigues Parameters.

tylerjereddy commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1012,14 +1015,110 @@ cdef class Rotation(object):
+        Notes
+        -----
+
+        .. versionadded:: 1.5.3
```
Just FYI -- this will probably have to be `1.6.0` if the feature is approved because we have already released the `1.5.x` series and almost always restrict point releases to bug fixes rather than enhancements/new features.

nbelakovski replied:
Updated

tylerjereddy commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1012,14 +1015,110 @@ cdef class Rotation(object):
+            quat[ind, 0] = 2 * cmrp[ind, 0] / (1 + ss)
+            quat[ind, 1] = 2 * cmrp[ind, 1] / (1 + ss)
+            quat[ind, 2] = 2 * cmrp[ind, 2] / (1 + ss)
+            quat[ind, 3] = (1 - ss) / (1 + ss)
```
minor: not sure if the compiler will optimize the repeated `(1 + ss)` calls or if assigning the value to i.e., a variable once might help

nbelakovski replied:
I'm not sure either, I went ahead and changed `ss` to `1+ss`

tylerjereddy commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1353,6 +1452,81 @@ cdef class Rotation(object):
+        Notes
+        -----
+
+        .. versionadded:: 1.5.3
```
`1.6.0`

nbelakovski commented:
> I see the manuscript has been cited > 2000 times, which may help for justification of breadth of interest, but I'm not immediately familiar with this myself.

I'm not sure myself what the breadth of interest is. I'd never heard of them until I started taking a course on Coursera for spacecraft attitude control. Perhaps it would be better to split this work out into it's own little Python library and then merge it into scipy if it seems popular enough?

nmayorov commented:
MRP fits well into Rotation, it's a small incremental addition which I support. I will review on the weekend.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -315,6 +315,99 @@ def test_rotvec_calc_pipeline():
+
+
+def test_from_generic_mrp():
+    mrp = [
```
Ugly formatting (and more cases below).

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -315,6 +315,99 @@ def test_rotvec_calc_pipeline():
+        [0, 0, 0, 1]
+        ])
+    assert_array_almost_equal(
+            Rotation.from_mrp(mrp).as_quat(),
```
Questionable formatting here as well.

nbelakovski replied:
Resolved, along with others. Do you guy prefer to hit the "resolve conversation" button when a comment is addressed or just leave it?

nbelakovski commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,102 @@ cdef class Rotation(object):
+        >>> r.as_rotvec().shape
+        (2, 3)
+
+        It is also possible to have a stack of a single rotaton:
```
Yes, but all the other definitions have this same example.

nmayorov replied:
The examples in `Rotation` are not very good, no point following them in my opinion.

nbelakovski commented:
I've updated the PR to address comments. In the update I have

- Made the definition more substantial and made it consistent across `from_mrp` and `as_mrp` with one small difference in `as_mrp` to denote that the rotation returned is <=180
- Modified the examples to use unity MRP and Euler angles where it made sense
-Renamed the `ss` variable. Renaming it to `denom` wasn't the best choice since it is also used in the numerator for calculating the scalar part of the quat, so I just went explicit and named it "mrp_squared_plus_1"
-Modified tests to have more consistent formatting (not sure if I got what you were after, but it looks more consistent to me now) and use the unity mrp in some cases
-Minor fixes (corrected the shape issue pointed out by @nmayorov , dropped use of the word "shadow" and updated docs accordingly, removed `quat` in the `as_mrp` calculation and just used a sign variable to make sure the set of MRPs being calculated corresponded to a rotation of <=180)

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,102 @@ cdef class Rotation(object):
+    def from_mrp(cls, mrp):
+        """Initialize from Modified Rodrigues Parameters (MRPs).
+
+        MRPs are a 3 element attitude description commonly used in the aerospace
```
description -> representation?

Do you think it's valuable to mention its usage in the aerospace industry? I'd say all common representations are widely used in the aerospace industry. I would probably suggest removing.

nbelakovski replied:
I went ahead and used your provided definition. I think MRPs are not widely used outside of aerospace, so I guess I'd say the target audience for that factoid is people outside of the aero industry, but I don't think it hurts one way or another if the factoid is in the docs or not, we can leave it out.

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,106 @@ cdef class Rotation(object):
+        """Initialize from Modified Rodrigues Parameters (MRPs).
+
+        MRPs are a 3 element attitude description commonly used in the aerospace
+        industry. It is similar to the rotation vector, except instead of scaling
```
Maybe just following definition in `from_rotvec` is a good idea.

MRPs is a 3 dimensional vector co-directional to the axis of rotation and whose magnitude is equal to ``tan(theta / 4)``, where ``theta`` is the angle of rotation (in radians) [1].

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,106 @@ cdef class Rotation(object):
+        the rotation vector by the angle of rotation, is it scaled by tan(theta/4).
+
+        MRPs have a singuarity at 360 degrees which can be easily avoided by ensuring
+        that the angle of rotation is always <= 180
```
`is always <= 180` -> `is always less or equal 180 degrees.`?

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,106 @@ cdef class Rotation(object):
+        -------
+        rotation : `Rotation` instance
+            Object containing the rotations represented by input rotation
+            vectors.
```
Not rotation vectors.

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,106 @@ cdef class Rotation(object):
+        Initialize a single rotation:
+
+        >>> r = R.from_mrp([0, 0, 1])
+        >>> r.as_euler('xyz')
```
Add `degrees=True` or use `as_rotvec` otherwise.

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,106 @@ cdef class Rotation(object):
+        >>> r.as_euler('xyz').shape
+        (2, 3)
+
+        It is also possible to have a stack of a single rotation:
```
I saw your consistency argument.

Not sure, I'm fine with either keeping or removing it.

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1357,6 +1460,80 @@ cdef class Rotation(object):
+            sign = -1 if self._quat[ind, 3] < 0 else 1
+
+            for i in range(3):
+                mrps[ind, i] = sign * self._quat[ind, i] / (1 + sign * self._quat[ind, 3])
```
Moving `cdef double denominator = 1 + sign * self._quat[ind, 3]` outside of the loop is a reasonable option.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -315,6 +315,99 @@ def test_rotvec_calc_pipeline():
+        [1, 2, 2],
+        [1, -1, 0.5],
+        [0, 0, 0]
+        ])
```
Super minor, but I'd put `])` on the previous line.

nbelakovski commented:
New update with latest feedback:

- Used @nmayorov's provided definition for MRPs and cleaned up the language a little
- Set Euler examples to use degrees
- Moved denominator calculation outside of for loop in `as_mrp`
- Other minor fixes (fixed copy/paste error spotted by @nmayorov, fixed formatting in test_rotation)

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,105 @@ cdef class Rotation(object):
+        """Initialize from Modified Rodrigues Parameters (MRPs).
+
+        MRPs is a 3 dimensional vector co-directional to the axis of rotation and whose
+        magnitude is equal to tan(theta / 4), where theta is the angle of rotation (in radians) [1].
```
I suggest adding double ticks (``) around "tan(theta / 4)" and "theta" to render it as code/expression.

The reference should be done as [1]_.

nbelakovski replied:
Done

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1016,6 +1019,105 @@ cdef class Rotation(object):
+        MRPs is a 3 dimensional vector co-directional to the axis of rotation and whose
+        magnitude is equal to tan(theta / 4), where theta is the angle of rotation (in radians) [1].
+
+        MRPs have a singuarity at 360 degrees which can be avoided by ensuring the angle of
```
This paragraph is written very solid now.

Optionally have -> has, because before we referred to MRPs as a singular entity (which is OK to my understanding). But I'm not super strong in grammar anyhow.

nbelakovski replied:
Good point, I actually changed "is" to "are" in the above paragraphs. "MRPs" is plural, so I think it makes sense to treat it as such. There are 3 of them after all...

nmayorov commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -1357,6 +1459,82 @@ cdef class Rotation(object):
+            # to a rotation of <= 180
+            sign = -1 if self._quat[ind, 3] < 0 else 1
+
+            denominator = 1 + sign * self._quat[ind, 3]
```
I suggest adding a test which will show that a returned MRPs will be correct and have magnitude <= 180 degrees, when the rotation was created with magnitude >= 180 degrees (through Euler or rotation vector). The scalar quaternion component must be negative in this case.

Reasoning - to test that this mathematical transformation does in fact work, because it's not obvious.

nbelakovski replied:
I think that's a good idea, test added

nbelakovski commented:
My latest changes seem to be causing some Windows failures, but I'm having trouble figuring out exactly what's going on. Any help?

tylerjereddy commented:
> My latest changes seem to be causing some Windows failures, but I'm having trouble figuring out exactly what's going on. Any help?

Not related to your changes, I should be able to fix shortly

nmayorov commented:
It looks very good now, I'm merging.

Thank you for the contribution!

nbelakovski commented:
> It looks very good now, I'm merging.
> 
> Thank you for the contribution!

Great working with you, looking forward to future contributions!

ilayn commented:
Thanks @nbelakovski . Please have a moment to boast about this in the release notes ;) 

https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0
